File: cz_cli/guide_builder.py
"""Build dynamic AI guide payload and generated skill documents."""

from __future__ import annotations

import copy
import difflib
import json
import os
from pathlib import Path
from typing import Any

import click

from cz_cli.version import __version__

GUIDE_GENERATOR_VERSION = "1.0.0"
DEFAULT_AI_GUIDE_BUDGET_CHARS = 20000

SKILL_TEMPLATE_PATH = Path(__file__).resolve().parent / "SKILL.template.md"
SKILL_OUTPUT_PATH = Path(__file__).resolve().parent / "skills" / "cz-cli" / "SKILL.md"


def _normalize_whitespace(value: str | None) -> str:
    if not value:
        return ""
    return " ".join(value.split())


def _as_click_group(command: click.Command) -> click.Group | None:
    if isinstance(command, click.Group):
        return command
    return None


def _safe_default(option: click.Option, ctx: click.Context) -> Any:
    try:
        default = option.get_default(ctx)
    except Exception:
        default = option.default

    if default in (None, "", (), []):
        return None

    if default.__class__.__name__ == "Sentinel":
        return None

    if isinstance(default, tuple):
        return list(default)

    return _json_safe_value(default)


def _json_safe_value(value: Any) -> Any:
    if value.__class__.__name__ == "Sentinel":
        return None

    if isinstance(value, tuple):
        return [_json_safe_value(item) for item in value]

    if isinstance(value, list):
        return [_json_safe_value(item) for item in value]

    if isinstance(value, dict):
        return {str(key): _json_safe_value(item) for key, item in value.items()}

    if isinstance(value, (str, int, float, bool)) or value is None:
        return value

    try:
        json.dumps(value)
    except TypeError:
        return str(value)
    return value


def _serialize_option(option: click.Option, ctx: click.Context) -> dict[str, Any]:
    names = [*option.opts, *option.secondary_opts]
    item: dict[str, Any] = {
        "flags": names,
        "required": bool(option.required),
        "takes_value": not option.is_flag,
    }

    metavar = _normalize_whitespace(option.make_metavar(ctx))
    if metavar and not option.is_flag:
        item["metavar"] = metavar

    help_text = _normalize_whitespace(option.help)
    if help_text:
        item["help"] = help_text

    default = _safe_default(option, ctx)
    if default is not None:
        item["default"] = default

    return item


def _serialize_argument(argument: click.Argument, ctx: click.Context) -> dict[str, Any]:
    item: dict[str, Any] = {
        "name": argument.human_readable_name,
        "required": bool(argument.required),
        "nargs": argument.nargs,
    }

    metavar = _normalize_whitespace(argument.make_metavar(ctx))
    if metavar:
        item["metavar"] = metavar

    default = argument.default
    if default not in (None, "", (), []):
        json_safe_default = _json_safe_value(default)
        if json_safe_default is not None:
            item["default"] = json_safe_default

    return item


def _extract_usage(command: click.Command, ctx: click.Context) -> str:
    raw = command.get_usage(ctx)
    if raw.startswith("Usage:"):
        raw = raw.split(":", 1)[1]
    return _normalize_whitespace(raw)


def _extract_description(command: click.Command) -> str:
    description = _normalize_whitespace(command.short_help)
    if description:
        return description
    description = _normalize_whitespace(command.help)
    if description:
        return description
    return "No description."


def _command_kind(command: click.Command) -> str:
    return "group" if _as_click_group(command) is not None else "command"


def _extract_shell_quote_hint(arguments: list[dict[str, Any]]) -> str | None:
    names = {str(arg.get("name", "")).upper() for arg in arguments}
    if "DDL" in names:
        return 'Shell tip: wrap DDL in quotes (e.g. "CREATE TABLE ...") or use --from-file.'
    if "STATEMENT" in names:
        return "Shell tip: wrap SQL statement in quotes, or pass -f/--file for complex statements."
    return None


def _build_command_entry(
    path: list[str], command: click.Command, ctx: click.Context
) -> dict[str, Any]:
    options: list[dict[str, Any]] = []
    arguments: list[dict[str, Any]] = []

    for param in command.params:
        if isinstance(param, click.Option):
            options.append(_serialize_option(param, ctx))
        elif isinstance(param, click.Argument):
            arguments.append(_serialize_argument(param, ctx))

    entry = {
        "name": " ".join(path),
        "kind": _command_kind(command),
        "usage": _extract_usage(command, ctx),
        "description": _extract_description(command),
        "options": options,
        "arguments": arguments,
    }
    shell_quote_hint = _extract_shell_quote_hint(arguments)
    if shell_quote_hint:
        entry["shell_quote_hint"] = shell_quote_hint
    return entry


def _walk_command_tree(
    command: click.Command,
    ctx: click.Context,
    path: list[str],
    inventory: list[dict[str, Any]],
) -> None:
    group = _as_click_group(command)
    if group is None:
        inventory.append(_build_command_entry(path, command, ctx))
        return

    # Include group commands themselves so every subcommand path has a usage signature.
    inventory.append(_build_command_entry(path, command, ctx))

    subcommand_names = sorted(group.list_commands(ctx))
    if not subcommand_names:
        return

    for name in subcommand_names:
        subcommand = group.get_command(ctx, name)
        if subcommand is None:
            continue
        child_ctx = click.Context(subcommand, info_name=name, parent=ctx)
        _walk_command_tree(subcommand, child_ctx, [*path, name], inventory)


def build_command_inventory(
    root: click.Command, *, cli_name: str = "cz-cli"
) -> list[dict[str, Any]]:
    """Return deterministic command metadata extracted from Click command tree."""
    inventory: list[dict[str, Any]] = []
    root_ctx = click.Context(root, info_name=cli_name)

    root_group = _as_click_group(root)
    if root_group is None:
        return [_build_command_entry([cli_name], root, root_ctx)]

    for top_name in sorted(root_group.list_commands(root_ctx)):
        top_command = root_group.get_command(root_ctx, top_name)
        if top_command is None:
            continue
        child_ctx = click.Context(top_command, info_name=top_name, parent=root_ctx)
        _walk_command_tree(top_command, child_ctx, [top_name], inventory)

    inventory.sort(key=lambda item: item["name"])
    return inventory


def _serialize_payload_length(payload: dict[str, Any]) -> int:
    return len(json.dumps(payload, ensure_ascii=False, indent=2))


def _trim_option_help_and_default(payload: dict[str, Any]) -> bool:
    changed = False
    for command in payload.get("commands", []):
        for option in command.get("options", []):
            if "help" in option:
                option.pop("help", None)
                changed = True
            if "default" in option:
                option.pop("default", None)
                changed = True
    return changed


def _drop_parameter_details(payload: dict[str, Any]) -> bool:
    changed = False
    for command in payload.get("commands", []):
        if command.pop("options", None) is not None:
            changed = True
        if command.pop("arguments", None) is not None:
            changed = True
    return changed


def _drop_global_examples(payload: dict[str, Any]) -> bool:
    global_options = payload.get("global_options", {})
    if "examples" not in global_options:
        return False
    global_options.pop("examples", None)
    return True


def _drop_recommended_workflow(payload: dict[str, Any]) -> bool:
    if "recommended_workflow" not in payload:
        return False
    payload.pop("recommended_workflow", None)
    return True


def _drop_tips(payload: dict[str, Any]) -> bool:
    if "tips" not in payload:
        return False
    payload.pop("tips", None)
    return True


def _drop_descriptions(payload: dict[str, Any]) -> bool:
    changed = False
    for command in payload.get("commands", []):
        if command.pop("description", None) is not None:
            changed = True
    return changed


def _apply_budget(payload: dict[str, Any], budget_chars: int) -> dict[str, Any]:
    before_chars = _serialize_payload_length(payload)
    applied_steps: list[str] = []

    if before_chars <= budget_chars:
        payload["truncation"] = {
            "applied": False,
            "budget_chars": budget_chars,
            "estimated_chars_before": before_chars,
            "estimated_chars_after": before_chars,
            "steps_applied": [],
            "mandatory_sections": [
                "global_options.usage_pattern",
                "commands[].name",
                "commands[].kind",
                "commands[].usage",
                "safety",
                "output_format",
                "exit_codes",
            ],
        }
        return payload

    trim_steps: list[tuple[str, Any]] = [
        ("drop_option_help_and_default", _trim_option_help_and_default),
        ("drop_parameter_details", _drop_parameter_details),
        ("drop_global_examples", _drop_global_examples),
        ("drop_recommended_workflow", _drop_recommended_workflow),
        ("drop_tips", _drop_tips),
        ("drop_command_descriptions", _drop_descriptions),
    ]

    for step_name, step_fn in trim_steps:
        changed = bool(step_fn(payload))
        current_chars = _serialize_payload_length(payload)
        if changed:
            applied_steps.append(step_name)
        if current_chars <= budget_chars:
            break

    after_chars = _serialize_payload_length(payload)
    payload["truncation"] = {
        "applied": True,
        "budget_chars": budget_chars,
        "estimated_chars_before": before_chars,
        "estimated_chars_after": after_chars,
        "steps_applied": applied_steps,
        "mandatory_sections": [
            "global_options.usage_pattern",
            "commands[].name",
            "commands[].kind",
            "commands[].usage",
            "safety",
            "output_format",
            "exit_codes",
        ],
    }
    return payload


def _resolve_budget(budget_chars: int | None = None, *, wide: bool = False) -> int:
    if budget_chars is not None:
        return max(2000, budget_chars)

    env_budget = os.environ.get("CZ_AI_GUIDE_BUDGET", "").strip()
    if not env_budget:
        return 200000 if wide else DEFAULT_AI_GUIDE_BUDGET_CHARS

    try:
        parsed = int(env_budget)
    except ValueError:
        return 200000 if wide else DEFAULT_AI_GUIDE_BUDGET_CHARS
    return max(2000, parsed)


def build_ai_guide(
    root: click.Command,
    *,
    cli_name: str = "cz-cli",
    cli_version: str = __version__,
    wide: bool = False,
    budget_chars: int | None = None,
) -> dict[str, Any]:
    """Build ai-guide payload from Click metadata with deterministic budget trimming."""
    command_inventory = build_command_inventory(root, cli_name=cli_name)

    root_ctx = click.Context(root, info_name=cli_name)
    global_options = [
        _serialize_option(option, root_ctx)
        for option in root.params
        if isinstance(option, click.Option)
    ]

    payload: dict[str, Any] = {
        "name": cli_name,
        "version": cli_version,
        "guide_generator_version": GUIDE_GENERATOR_VERSION,
        "description": "AI-Agent-friendly CLI for ClickZetta Lakehouse",
        "global_options": {
            "order": "Global options must appear before subcommand.",
            "usage_pattern": (
                f"{cli_name} [GLOBAL_OPTIONS] <subcommand> [args] "
                f"(or run {cli_name} <subcommand> [args] --output pretty)"
            ),
            "entry_commands": ["cz-cli", "clickzetta-cli"],
            "options": global_options,
            "examples": [
                f'{cli_name} --profile dev sql "SELECT 1"',
                f"{cli_name} --output table task list --limit 5",
                f"{cli_name} runs list --task my_task --run-type REFILL --limit 1",
            ],
        },
        "recommended_workflow": [
            f"{cli_name} profile create <name> --username ... --password ... --instance ... --workspace ...",
            f"{cli_name} schema list",
            f"{cli_name} table list --schema public",
            f'{cli_name} sql "SELECT ... LIMIT 20"',
        ],
        "commands": command_inventory,
        "output_format": {
            "success": {"ok": True, "data": "...", "time_ms": "N"},
            "error": {"ok": False, "error": {"code": "...", "message": "..."}},
            "note": "Use -o pretty for colorized human-friendly JSON output.",
        },
        "safety": {
            "write_protection": "Write operations require --write. DELETE/UPDATE without WHERE are blocked.",
            "confirmation": "task online/offline, runs stop/refill, executions stop require confirmation unless -y.",
            "pagination": "task list / runs list / executions list default to page 1; use --page/--page-size or --limit.",
        },
        "exit_codes": {"0": "success", "1": "business error", "2": "usage error"},
        "tips": {
            "help": f"Run '{cli_name} <subcommand> --help' to inspect exact parameter contracts.",
            "profile": "Use --profile for reusable connection config; protocol can be overridden per profile.",
        },
    }

    # Default view is compact: keep command signatures and safety contracts,
    # and let caller opt into parameter-level details via --wide.
    if not wide:
        _drop_parameter_details(payload)

    return _apply_budget(copy.deepcopy(payload), _resolve_budget(budget_chars, wide=wide))


def _skill_inventory_markdown(inventory: list[dict[str, Any]]) -> str:
    grouped: dict[str, list[dict[str, Any]]] = {}
    for item in inventory:
        top_level = item["name"].split(" ", 1)[0]
        grouped.setdefault(top_level, []).append(item)

    lines: list[str] = []
    for top_level in sorted(grouped):
        lines.append(f"### `{top_level}`")
        for entry in sorted(grouped[top_level], key=lambda x: x["name"]):
            lines.append(f"- `{entry['usage']}` - {entry['description']}")
        lines.append("")
    return "\n".join(lines).strip() + "\n"


def render_skill_markdown(
    root: click.Command,
    *,
    template_text: str,
    cli_name: str = "cz-cli",
    cli_version: str = __version__,
) -> str:
    """Render skill markdown from template and dynamic command inventory."""
    inventory = build_command_inventory(root, cli_name=cli_name)
    command_inventory = _skill_inventory_markdown(inventory)

    rendered = template_text
    replacements = {
        "{{CLI_VERSION}}": cli_version,
        "{{GENERATOR_VERSION}}": GUIDE_GENERATOR_VERSION,
        "{{COMMAND_COUNT}}": str(len(inventory)),
        "{{COMMAND_INVENTORY}}": command_inventory.rstrip(),
    }
    for key, value in replacements.items():
        rendered = rendered.replace(key, value)

    return rendered.rstrip() + "\n"


def generate_skill_markdown(
    root: click.Command,
    *,
    template_path: Path = SKILL_TEMPLATE_PATH,
    cli_name: str = "cz-cli",
    cli_version: str = __version__,
) -> str:
    """Generate skill markdown text from template path."""
    template_text = template_path.read_text(encoding="utf-8")
    return render_skill_markdown(
        root,
        template_text=template_text,
        cli_name=cli_name,
        cli_version=cli_version,
    )


def write_generated_skill(
    root: click.Command,
    *,
    output_path: Path = SKILL_OUTPUT_PATH,
    template_path: Path = SKILL_TEMPLATE_PATH,
    cli_name: str = "cz-cli",
    cli_version: str = __version__,
) -> Path:
    """Generate and write skill markdown to disk."""
    rendered = generate_skill_markdown(
        root,
        template_path=template_path,
        cli_name=cli_name,
        cli_version=cli_version,
    )
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(rendered, encoding="utf-8")
    return output_path


def skill_drift_diff(
    root: click.Command,
    *,
    output_path: Path = SKILL_OUTPUT_PATH,
    template_path: Path = SKILL_TEMPLATE_PATH,
    cli_name: str = "cz-cli",
    cli_version: str = __version__,
) -> str:
    """Return unified diff when committed skill content drifts from generator output."""
    expected = generate_skill_markdown(
        root,
        template_path=template_path,
        cli_name=cli_name,
        cli_version=cli_version,
    )
    actual = output_path.read_text(encoding="utf-8") if output_path.exists() else ""
    if expected == actual:
        return ""

    diff = difflib.unified_diff(
        actual.splitlines(),
        expected.splitlines(),
        fromfile=str(output_path),
        tofile=f"{output_path} (generated)",
        lineterm="",
    )
    return "\n".join(diff)

File: packages/cz-cli/src/guide-builder.ts
import { VERSION } from "./version.js"

const GUIDE_GENERATOR_VERSION = "1.0.0"
const DEFAULT_BUDGET_CHARS = 40000

export interface CommandEntry {
  name: string
  kind: "command" | "group"
  description: string
  usage?: string
  options?: OptionEntry[]
  arguments?: ArgumentEntry[]
  shell_quote_hint?: string
  examples?: { cmd: string; desc: string }[]
}

interface OptionEntry {
  flags: string
  required: boolean
  takes_value: boolean
  help?: string
  default?: unknown
}

interface ArgumentEntry {
  name: string
  required: boolean
}

interface GuidePayload {
  name: string
  version: string
  guide_generator_version: string
  description: string
  global_options: Record<string, unknown>
  recommended_workflow: string[]
  commands: Record<string, unknown>[]
  output_format: Record<string, unknown>
  safety: Record<string, string>
  exit_codes: Record<string, string>
  tips: Record<string, string>
  truncation?: Record<string, unknown>
}

const commandRegistry: CommandEntry[] = []

export function registerCommand(entry: CommandEntry): void {
  commandRegistry.push(entry)
}

export function registerCommands(entries: CommandEntry[]): void {
  for (const e of entries) commandRegistry.push(e)
}

function serializePayloadLength(payload: Record<string, unknown>): number {
  return JSON.stringify(payload, null, 2).length
}

function dropOptionHelpAndDefault(payload: Record<string, unknown>): boolean {
  let changed = false
  for (const cmd of (payload.commands ?? []) as Record<string, unknown>[]) {
    for (const opt of (cmd.options ?? []) as Record<string, unknown>[]) {
      if ("help" in opt) { delete opt.help; changed = true }
      if ("default" in opt) { delete opt.default; changed = true }
    }
  }
  return changed
}

function dropParameterDetails(payload: Record<string, unknown>): boolean {
  let changed = false
  for (const cmd of (payload.commands ?? []) as Record<string, unknown>[]) {
    if (cmd.options) { delete cmd.options; changed = true }
    if (cmd.arguments) { delete cmd.arguments; changed = true }
  }
  return changed
}

function dropCommandExamples(payload: Record<string, unknown>): boolean {
  let changed = false
  for (const cmd of (payload.commands ?? []) as Record<string, unknown>[]) {
    if (cmd.examples) { delete cmd.examples; changed = true }
  }
  if (payload.command_examples) { delete payload.command_examples; changed = true }
  return changed
}

function dropGlobalExamples(payload: Record<string, unknown>): boolean {
  const go = payload.global_options as Record<string, unknown> | undefined
  if (!go || !("examples" in go)) return false
  delete go.examples
  return true
}

function dropRecommendedWorkflow(payload: Record<string, unknown>): boolean {
  if (!("recommended_workflow" in payload)) return false
  delete payload.recommended_workflow
  return true
}

function dropTips(payload: Record<string, unknown>): boolean {
  if (!("tips" in payload)) return false
  delete payload.tips
  return true
}

function dropDescriptions(payload: Record<string, unknown>): boolean {
  let changed = false
  for (const cmd of (payload.commands ?? []) as Record<string, unknown>[]) {
    if (cmd.description) { delete cmd.description; changed = true }
  }
  return changed
}

function applyBudget(payload: Record<string, unknown>, budgetChars: number): Record<string, unknown> {
  const beforeChars = serializePayloadLength(payload)
  const appliedSteps: string[] = []

  if (beforeChars <= budgetChars) {
    payload.truncation = {
      applied: false,
      budget_chars: budgetChars,
      estimated_chars_before: beforeChars,
      estimated_chars_after: beforeChars,
      steps_applied: [],
    }
    return payload
  }

  const steps: [string, (p: Record<string, unknown>) => boolean][] = [
    ["drop_option_help_and_default", dropOptionHelpAndDefault],
    ["drop_parameter_details", dropParameterDetails],
    ["drop_command_examples", dropCommandExamples],
    ["drop_global_examples", dropGlobalExamples],
    ["drop_recommended_workflow", dropRecommendedWorkflow],
    ["drop_tips", dropTips],
    ["drop_command_descriptions", dropDescriptions],
  ]

  for (const [name, fn] of steps) {
    if (fn(payload)) appliedSteps.push(name)
    if (serializePayloadLength(payload) <= budgetChars) break
  }

  payload.truncation = {
    applied: true,
    budget_chars: budgetChars,
    estimated_chars_before: beforeChars,
    estimated_chars_after: serializePayloadLength(payload),
    steps_applied: appliedSteps,
    mandatory_sections: ["global_options", "recommended_workflow", "output_format", "safety", "exit_codes"],
  }
  return payload
}

function resolveBudget(budgetChars: number | undefined, wide: boolean): number {
  if (budgetChars !== undefined) return Math.max(2000, budgetChars)
  const env = process.env.CZ_AI_GUIDE_BUDGET?.trim()
  if (env) {
    const parsed = parseInt(env, 10)
    if (!isNaN(parsed)) return Math.max(2000, parsed)
  }
  return wide ? 300000 : DEFAULT_BUDGET_CHARS
}

export function buildToonPayload(payload: Record<string, unknown>): Record<string, unknown> {
  const result = structuredClone(payload)
  const commandsFlat: Record<string, unknown>[] = []
  const commandExamples: Record<string, unknown> = {}
  const commandHints: Record<string, string> = {}

  for (const cmd of (result.commands ?? []) as Record<string, unknown>[]) {
    const flat: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(cmd)) {
      if (k !== "examples" && k !== "shell_quote_hint") flat[k] = v
    }
    commandsFlat.push(flat)
    if (cmd.examples) commandExamples[cmd.name as string] = cmd.examples
    if (cmd.shell_quote_hint) commandHints[cmd.name as string] = cmd.shell_quote_hint as string
  }

  result.commands = commandsFlat
  if (Object.keys(commandExamples).length > 0) result.command_examples = commandExamples
  if (Object.keys(commandHints).length > 0) result.command_hints = commandHints
  return result
}

export function buildAiGuide(options?: { wide?: boolean; budgetChars?: number }): GuidePayload {
  const wide = options?.wide ?? false
  const cliName = "cz-cli"

  const sorted = [...commandRegistry].sort((a, b) => a.name.localeCompare(b.name))
  const commandsDicts = sorted.map((e) => {
    const d: Record<string, unknown> = { name: e.name, kind: e.kind, description: e.description }
    if (e.usage) d.usage = e.usage
    if (e.options && e.options.length > 0) d.options = e.options
    if (e.arguments && e.arguments.length > 0) d.arguments = e.arguments
    if (e.shell_quote_hint) d.shell_quote_hint = e.shell_quote_hint
    if (e.examples && e.examples.length > 0) d.examples = e.examples
    return d
  })

  const payload: Record<string, unknown> = {
    name: cliName,
    version: VERSION,
    guide_generator_version: GUIDE_GENERATOR_VERSION,
    description: "AI-Agent-friendly CLI for ClickZetta Lakehouse",
    global_options: {
      order: "Global options must appear before subcommand.",
      usage_pattern: `${cliName} [GLOBAL_OPTIONS] <subcommand> [args] (or run ${cliName} <subcommand> [args] --format pretty)`,
      entry_commands: ["cz-cli", "clickzetta-cli"],
      options: [
        { flags: "--profile, -p", required: false, takes_value: true, help: "Profile name" },
        { flags: "--format", required: false, takes_value: true, help: "Output format", default: "json" },
        { flags: "--debug, -d", required: false, takes_value: false, help: "Debug mode" },
      ],
      examples: [
        `${cliName} --profile dev sql "SELECT 1"`,
        `${cliName} --format table task list --limit 5`,
        `${cliName} runs list --task my_task --run-type REFILL --limit 1`,
      ],
    },
    recommended_workflow: [
      `${cliName} profile create <name> --username ... --password ... --instance ... --workspace ...`,
      `${cliName} schema list`,
      `${cliName} table list --schema public`,
      `${cliName} sql "SELECT ... LIMIT 20"`,
    ],
    commands: commandsDicts,
    output_format: {
      success: { ok: true, data: "...", time_ms: "N" },
      error: { ok: false, error: { code: "...", message: "..." } },
      note: "Use --format pretty for colorized human-friendly JSON output.",
    },
    safety: {
      write_protection: "Write operations require --write. DELETE/UPDATE without WHERE are blocked.",
      confirmation: "task online/offline, runs stop/refill, executions stop require confirmation unless -y.",
      pagination: "task list / runs list / executions list default to page 1; use --page/--page-size or --limit.",
    },
    exit_codes: { "0": "success", "1": "business error", "2": "usage error" },
    tips: {
      help: `Run '${cliName} <subcommand> --help' to inspect exact parameter contracts.`,
      profile: "Use --profile for reusable connection config; protocol can be overridden per profile.",
    },
  }

  if (!wide) dropParameterDetails(payload)

  return applyBudget(structuredClone(payload), resolveBudget(options?.budgetChars, wide)) as unknown as GuidePayload
}

export function registerStaticCommands(): void {
  registerCommands([
    { name: "sql", kind: "command", description: "Execute SQL against ClickZetta",
      options: [
        { flags: "--write", required: false, takes_value: false, help: "Allow write operations" },
        { flags: "-e, --execute", required: false, takes_value: true, help: "SQL string" },
        { flags: "-f, --file", required: false, takes_value: true, help: "Read SQL from file" },
        { flags: "--sync", required: false, takes_value: false, help: "Wait for result" },
        { flags: "--no-limit", required: false, takes_value: false, help: "Disable LIMIT guard" },
        { flags: "-N, --no-header", required: false, takes_value: false, help: "Omit column headers" },
      ],
      arguments: [{ name: "statement", required: false }],
      shell_quote_hint: "Shell tip: wrap SQL statement in quotes, or pass -f/--file for complex statements.",
      examples: [
        { cmd: 'cz-cli sql "SELECT 1"', desc: "Run a simple query" },
        { cmd: 'cz-cli sql -e "SELECT * FROM t LIMIT 10" --sync', desc: "Synchronous query" },
        { cmd: "cz-cli sql -f query.sql --write", desc: "Execute write SQL from file" },
      ],
    },
    { name: "schema", kind: "group", description: "Manage schemas" },
    { name: "schema list", kind: "command", description: "List schemas" },
    { name: "schema describe", kind: "command", description: "Describe a schema",
      arguments: [{ name: "name", required: true }] },
    { name: "schema create", kind: "command", description: "Create a schema",
      arguments: [{ name: "name", required: true }] },
    { name: "schema drop", kind: "command", description: "Drop a schema",
      arguments: [{ name: "name", required: true }] },
    { name: "table", kind: "group", description: "Manage tables" },
    { name: "table list", kind: "command", description: "List tables" },
    { name: "table describe", kind: "command", description: "Describe a table",
      arguments: [{ name: "name", required: true }] },
    { name: "table preview", kind: "command", description: "Preview table data",
      arguments: [{ name: "name", required: true }] },
    { name: "table stats", kind: "command", description: "Get table row count",
      arguments: [{ name: "name", required: true }] },
    { name: "table history", kind: "command", description: "Show table history",
      arguments: [{ name: "name", required: false }] },
    { name: "table create", kind: "command", description: "Create a table from DDL",
      arguments: [{ name: "ddl", required: true }],
      shell_quote_hint: 'Shell tip: wrap DDL in quotes (e.g. "CREATE TABLE ...") or use --from-file.',
    },
    { name: "table drop", kind: "command", description: "Drop a table",
      arguments: [{ name: "name", required: true }] },
    { name: "workspace", kind: "group", description: "Manage workspaces" },
    { name: "workspace list", kind: "command", description: "List workspaces" },
    { name: "workspace current", kind: "command", description: "Show current workspace" },
    { name: "profile", kind: "group", description: "Manage connection profiles" },
    { name: "profile list", kind: "command", description: "List profiles" },
    { name: "profile detail", kind: "command", description: "Show profile details",
      arguments: [{ name: "name", required: true }] },
    { name: "profile create", kind: "command", description: "Create a profile",
      arguments: [{ name: "name", required: true }] },
    { name: "profile update", kind: "command", description: "Update a profile",
      arguments: [{ name: "name", required: true }] },
    { name: "profile delete", kind: "command", description: "Delete a profile",
      arguments: [{ name: "name", required: true }] },
    { name: "profile use", kind: "command", description: "Set default profile",
      arguments: [{ name: "name", required: true }] },
    { name: "task", kind: "group", description: "Manage Studio tasks" },
    { name: "task list", kind: "command", description: "List tasks" },
    { name: "task create", kind: "command", description: "Create a new task",
      arguments: [{ name: "name", required: true }] },
    { name: "task content", kind: "command", description: "Get task content",
      arguments: [{ name: "id", required: true }] },
    { name: "task save-content", kind: "command", description: "Save task script",
      arguments: [{ name: "id", required: true }] },
    { name: "task save-config", kind: "command", description: "Save task schedule config",
      arguments: [{ name: "id", required: true }] },
    { name: "task execute", kind: "command", description: "Execute a task ad-hoc",
      arguments: [{ name: "id", required: true }] },
    { name: "task online", kind: "command", description: "Publish a task",
      arguments: [{ name: "id", required: true }] },
    { name: "task offline", kind: "command", description: "Take a task offline",
      arguments: [{ name: "id", required: true }] },
    { name: "task deps", kind: "command", description: "Show task dependencies",
      arguments: [{ name: "id", required: true }] },
    { name: "runs", kind: "group", description: "Manage task run instances" },
    { name: "runs list", kind: "command", description: "List run instances" },
    { name: "runs detail", kind: "command", description: "Get run detail",
      arguments: [{ name: "id", required: true }] },
    { name: "runs wait", kind: "command", description: "Poll until run completes",
      arguments: [{ name: "id", required: true }] },
    { name: "runs logs", kind: "command", description: "Get run logs",
      arguments: [{ name: "id", required: true }] },
    { name: "runs deps", kind: "command", description: "View run dependencies",
      arguments: [{ name: "id", required: true }] },
    { name: "runs stop", kind: "command", description: "Stop a running instance",
      arguments: [{ name: "id", required: true }] },
    { name: "runs rerun", kind: "command", description: "Rerun a failed instance",
      arguments: [{ name: "id", required: true }] },
    { name: "runs refill", kind: "command", description: "Submit backfill job" },
    { name: "runs stats", kind: "command", description: "Get run statistics summary" },
    { name: "attempts", kind: "group", description: "Manage attempt records" },
    { name: "attempts list", kind: "command", description: "List attempts for a run" },
    { name: "attempts log", kind: "command", description: "Get attempt log" },
    { name: "job", kind: "group", description: "Job performance tools" },
    { name: "job status", kind: "command", description: "Check status/summary of a SQL job",
      arguments: [{ name: "job-id", required: true }] },
    { name: "job result", kind: "command", description: "Fetch result set of a SQL job",
      arguments: [{ name: "job-id", required: true }] },
    { name: "job profile", kind: "command", description: "Show flattened job profile basics from getJobProfile",
      arguments: [{ name: "job-id", required: true }],
      options: [
        { flags: "--raw", required: false, takes_value: false, help: "Show raw profile content" },
        { flags: "--limit", required: false, takes_value: false, help: "Limit raw profile output; use --no-limit to show the full payload" },
        { flags: "--path <file>", required: false, takes_value: true, help: "Write the full raw profile payload to a file" },
      ] },
  ])
}

/**
 * Return the sorted command inventory. Ensures static commands are registered first.
 */
export function buildCommandInventory(): CommandEntry[] {
  if (commandRegistry.length === 0) registerStaticCommands()
  return [...commandRegistry].sort((a, b) => a.name.localeCompare(b.name))
}
